clients/drcachesim/tracer/tracer.cpp

```diff
@@ -413,13 +427,26 @@ memtrace(void *drcontext, bool skip_size_cap)
                                               dr_get_thread_id(drcontext));
     pipe_start = data->buf_base;
     pipe_end = pipe_start;
-    if (!skip_size_cap && op_max_trace_size.get_value() > 0 &&
-        data->bytes_written > op_max_trace_size.get_value()) {
+    if (!skip_size_cap &&
+        (is_bytes_written_beyond_trace_max(data) || is_num_refs_beyond_global_max())) {
         /* We don't guarantee to match the limit exactly so we allow one buffer
          * beyond.  We also don't put much effort into reducing overhead once
          * beyond the limit: we still instrument and come here.
          */
         do_write = false;
+        if (is_num_refs_beyond_global_max()) {
+            /* std::atomic *should* be safe (we can assert std::atomic_is_lock_free())
+             * but to avoid any risk we use DR's atomics.  The store cannot be moved
+             * above the load due to the acquire-release semantics.
+             */
+            static int notify_once;
+            if (dr_atomic_load32(&notify_once) == 0) {
+                int count = dr_atomic_add32_return_sum(&notify_once, 1);
+                if (count == 1) {
+                    NOTIFY(0, "Hit -max_global_trace_refs: disabling tracing.\n");
```

Comment:
As `num_refs_racy` is reset on re-attach, I'm assuming that the trace refs limit is per-attach - right? Maybe add this to the option string.

So, if we reach the limit again after re-attaching, do we want to also NOTIFY again?

Reply:
*Everything* is per attach.  It doesn't make sense to me to add to every single option that it will start over on a re-attach.

What is missing is re-setting the notify_once on detach.  Done.

Reply:
Makes sense.